### PR TITLE
pushing discovered disk to etcd

### PR DESCRIPTION
**Pull request by pawanpraka1**
```
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: disks.openebs.io
spec:
  group: openebs.io
  version: v1alpha1
  scope: Cluster
  names:
    plural: disks
    singular: disk
    kind: Disk
    shortNames:
    - disk

```
```
$ ndm start --kubeconfig=$HOME/.kube/config
E0329 00:40:15.252388   29860 ctrl_main.go:36] failed to get k8s Incluster config. unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined
I0329 00:40:15.258952   29860 ctrl.go:61] Started the controller
I0329 00:40:15.397065   29860 ctrl.go:125] disk object already exist : sda
I0329 00:40:15.400448   29860 ctrl.go:127] pushing disk object : sdb
I0329 00:40:15.406694   29860 ctrl.go:80] Created disk object : disk-b3f828db7cd829ac237e90ccf433bd17
```

```
$ kubectl get disk
NAME                                    AGE
disk-0x500003949d304e06                 45s
disk-b3f828db7cd829ac237e90ccf433bd17   5s
```
```
$ kubectl describe disks
Name:         disk-0x500003949d304e06
Namespace:    
Labels:       <none>
Annotations:  <none>
API Version:  openebs.io/v1alpha1
Kind:         Disk
Metadata:
  Cluster Name:                   
  Creation Timestamp:             2018-03-28T19:09:35Z
  Deletion Grace Period Seconds:  <nil>
  Deletion Timestamp:             <nil>
  Initializers:                   <nil>
  Resource Version:               2150111
  Self Link:                      /apis/openebs.io/v1alpha1/disk-0x500003949d304e06
  UID:                            8e008a5b-32bb-11e8-bab3-e4115b2f5dfc
Spec:
  Capacity:
    Storage:  500107862016
  Details:
    Model:   TOSHIBA MK5061GS
    Serial:  23FCYE2IF
    Vendor:  ATA     
  Path:      /dev/sda
Events:      <none>


Name:         disk-b3f828db7cd829ac237e90ccf433bd17
Namespace:    
Labels:       <none>
Annotations:  <none>
API Version:  openebs.io/v1alpha1
Kind:         Disk
Metadata:
  Cluster Name:                   
  Creation Timestamp:             2018-03-28T19:10:15Z
  Deletion Grace Period Seconds:  <nil>
  Deletion Timestamp:             <nil>
  Initializers:                   <nil>
  Resource Version:               2150156
  Self Link:                      /apis/openebs.io/v1alpha1/disk-b3f828db7cd829ac237e90ccf433bd17
  UID:                            a5a36f0e-32bb-11e8-bab3-e4115b2f5dfc
Spec:
  Capacity:
    Storage:  15597568000
  Details:
    Model:   Ultra           
    Serial:  SanDisk.Ultra...........4C531001520618104171
    Vendor:  SanDisk 
  Path:      /dev/sdb
Events:      <none>
```

### CODE REVIEW

File: pkg/apis/ndm/v1alpha1/types.go [hunk 1]
```diff
@@ -71,3 +71,37 @@ type StoragePoolList struct {
...
+type DiskObject struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata, omitempty"`
+
+	Spec DiskObjectSpec `json:"spec"`
+}
+
+// DiskObjectSpec is the specification for the disk stored as CRD
+type DiskObjectSpec struct {
+	Uuid       string `json:"uuid"`  //unique identifier for the disk
```
Comment (gila) [inline pkg/apis/ndm/v1alpha1/types.go]: How is this uuid determined? We should use GUID of the disk, they are designed for that and it allows us to determine that disks are moved (hot swapped) between nodes as well.  I'm referring to the NAA IEEE number see SPC4 section 7.

File: pkg/apis/ndm/v1alpha1/types.go [hunk 2]
```diff
@@ -71,3 +71,37 @@ type StoragePoolList struct {
...
+	Name       string `json:"name"`  //disk name (e.g. sdb, sdc)
+	Path       string `json:"path"`  //disk path (e.g. /dev/sdb)
```
Comment (gila) [inline pkg/apis/ndm/v1alpha1/types.go]: This should be an array -- as there is disk/by-id disk/by-guid (hey?! our guid=)) etc.

File: pkg/apis/ndm/v1alpha1/types.go [hunk 3]
```diff
@@ -71,3 +71,37 @@ type StoragePoolList struct {
...
+	Size       string `json:"size"`  //disk size (e.g. 200G)
```
Comment (gila) [inline pkg/apis/ndm/v1alpha1/types.go]: Its better to put this number in bytes such that UI's and what not, don't have to parse our strings and can decide how they want to display the numbers themselves.

**Comment by kmova** [comment]
@pawanpraka1 : Few comments to keep this in sync with the design: 
- Can you modify the DiskObjects to Disk Resources as mentioned in this example : https://github.com/openebs/node-disk-manager/pull/1/files#diff-33325f0c9f4b7b398ce37179dc3e6db5
- The types should reflect the example or vice versa: https://github.com/openebs/node-disk-manager/pull/1/files#diff-af33dcbe18ed6c8fbd2db0825dbf6b8a

File: pkg/apis/ndm/v1alpha1/types.go [hunk 4]
```diff
@@ -71,3 +71,47 @@ type StoragePoolList struct {
...
+// DiskSpec is the specification for the disk stored as CRD
+type DiskSpec struct {
+	Uuid       string `json:"uuid"`         //unique identifier for the disk
+	Path       string `json:"path"`         //disk path (e.g. /dev/sdb)
+	Capacity DiskCapacity `json:"capacity"` //capacity (e.g. size, used)
+	Details DiskDetails `json:"details"`    //disk details (e.g. model, serial)
+}
+
+type DiskCapacity struct {
+	Storage   uint64 `json:"storage"`       //disk size (e.g. 200G)
```
Comment (gila) [inline pkg/apis/ndm/v1alpha1/types.go]: The comment should also be updated, secondly if we name it bytes it makes more sense perhaps? i.e DiskCapacity.bytes = XXX ? and then we can even make it DiskCapacity.HumanReadble = 20G or whatever in later versions of the API?

File: pkg/apis/ndm/v1alpha1/types.go [hunk 5]
```diff
@@ -71,3 +71,47 @@ type StoragePoolList struct {
...
+type Disk struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata, omitempty"`
+
+	Spec DiskSpec `json:"spec"`
+}
+
+// DiskSpec is the specification for the disk stored as CRD
+type DiskSpec struct {
+	Uuid       string `json:"uuid"`         //unique identifier for the disk
```
Comment (gila) [inline pkg/apis/ndm/v1alpha1/types.go]: Whats this UUID based on? can we use the disk GUID SPC4, section 7.

File: pkg/apis/ndm/v1alpha1/types.go [hunk 6]
```diff
@@ -71,3 +71,47 @@ type StoragePoolList struct {
...
+	Path       string `json:"path"`         //disk path (e.g. /dev/sdb)
```
Comment (gila) [inline pkg/apis/ndm/v1alpha1/types.go]: Should be array

File: cmd/controller/ctrl.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+	"github.com/golang/glog"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	apis "github.com/openebs/node-disk-manager/pkg/apis/ndm/v1alpha1"
+	clientset "github.com/openebs/node-disk-manager/pkg/client/clientset/versioned"
+)
+
+const (
+	NdmKind = "Disk"
```
Comment (AmitKumarDas) [inline cmd/controller/ctrl.go]: The var name can be NDMKind

File: cmd/controller/ctrl.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+	NdmVersion = "openebs.io/v1alpha1"
```
Comment (AmitKumarDas) [inline cmd/controller/ctrl.go]: NDMVersion

File: cmd/controller/ctrl.go [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+	if err != true {
+		glog.Error("Unable to push disk objects to etcd ", err)
+	}
+	<-stopCh
+	glog.Info("Shutting down the controller")
+
+	return nil
+}
+
+func CreateDo(dr apis.Disk, c *Controller) {
```
Comment (AmitKumarDas) [inline cmd/controller/ctrl.go]: CreateDO

Comment (AmitKumarDas) [reply cmd/controller/ctrl.go]: Similarly other function nam3s. Also check if they can be private vs. public functions

File: cmd/controller/ctrl.go (hunk 7 above)
Comment (pawanpraka1) [reply cmd/controller/ctrl.go]: yeah

File: cmd/controller/ctrl.go (hunk 8 above)
Comment (pawanpraka1) [reply cmd/controller/ctrl.go]: done

File: cmd/controller/ctrl.go (hunk 9 above)
Comment (pawanpraka1) [reply cmd/controller/ctrl.go]: yeah, they can be private function. Chaged them.

**Comment by pawanpraka1** [comment]
@kmova, made the spec in sync with the design doc.

File: pkg/apis/ndm/v1alpha1/types.go (hunk 1 above)
Comment (pawanpraka1) [reply pkg/apis/ndm/v1alpha1/types.go]: we can discuss this on the design. Please refer https://github.com/openebs/node-disk-manager/pull/1

File: pkg/apis/ndm/v1alpha1/types.go (hunk 5 above)
Comment (pawanpraka1) [reply pkg/apis/ndm/v1alpha1/types.go]: we can discuss this on the design. Please refer #1

**Comment by kmova** [comment]
The review comments related to uuid, and other attributes will be addressed via design discussion. This PR is merged to have the code (disk objects) available for building the metrics/monitoring capabilities.

**Comment by shovanmaity** [comment]
fixes - https://github.com/openebs/node-disk-manager/issues/10

**Pull request merged**